### PR TITLE
feat: add select uninitialized participants on internal criteria #11

### PR DESCRIPTION
**Pull request by hubertusputu1**
## Context 
#11 

## Features
- [x] able to select uninitialized participants from imported participants on internal DB

### CODE REVIEW

File: internal/mturk/step.go [hunk 1]
```diff
@@ -465,10 +465,18 @@ func (s *Session) endMTurkHITStep(ctx context.Context, project *ent.Project, run
...
 					AddProjects(project).
 					AddSteps(stepRun).
-					SetUninitialized(false).
 					Save(ctx)
 				if err != nil {
 					log.Error().Msgf("could not update participant with workerID %s for stepRun %s", *assignment.WorkerId, stepRun.ID)
+				}
+			}
+
+			if *p.Uninitialized != false {
```
Comment (npaton) [inline internal/mturk/step.go:474]: You're dereferencing a pointer without checking if it's null first: `p.Uninitialized != null && *p.Uninitialized != false`.

File: internal/mturk/step.go [hunk 2]
```diff
@@ -465,10 +465,18 @@ func (s *Session) endMTurkHITStep(ctx context.Context, project *ent.Project, run
...
+				_, err = p.Update().
+					SetUninitialized(false).
+					Save(ctx)
+				if err != nil {
+					log.Error().Msgf("could not set participant unitialized with workerID %s for stepRun ", *assignment.WorkerId, stepRun.ID)
```
Comment (npaton) [inline internal/mturk/step.go:479]: uninitialized

File: internal/runtime/process.go [hunk 3]
```diff
@@ -255,12 +255,17 @@ func (r *runState) filterParticipants(ctx context.Context, tx *ent.Tx, limit int
...
 
 	initializedParticipants := make([]*ent.Participant, 0)
-	for _, participant := range participants {
-		if participant.Uninitialized != nil && *participant.Uninitialized == true {
-			continue
-		}
 
-		initializedParticipants = append(initializedParticipants, participant)
+	if uninitialized == true {
+		initializedParticipants = participants
```
Comment (npaton) [inline internal/runtime/process.go:260]: This should actually be the opposite of below (could be one loop I think), we only want uninitialized participants if true. The idea is we want to select only uninitialized participants to make them go through a HIT so they are now initialized, so we want to be able to isolate them in both directions (only initialized / only uninitialized).

File: web/src/components/templates/Template.svelte [hunk 4]
```diff
@@ -146,13 +147,25 @@
...
+        options={selectionTypes}
+        placeholder="Particpant Selection Type" />
+    </div>
+    {#if template.selectionType === 'INTERNAL_DB'}
+      <div class="flex flex-column pt-5 max-w-xs">
+        <Checkbox bind:checked={template.internalCriteria.uninitialized}>
+          <Label
+            forID="unitializedParticipants"
+            text="Select Uninitialized Participants"
+            question="Uninitialized Participants are Participants that have been imported, and do not yet have a HIT associated with them, so we cannot send messages or bonuses to them" />
```
Comment (npaton) [inline web/src/components/templates/Template.svelte:165]: I guess given my previous comment, we should add some details here. At the end of the description, you can add: 

"Uninitialized participants are *not* included in an Internal DB selection. On the contrary, if you check this box, *only* uninitialized participants will be selected."

Hum, that's a bit of a mouthful... Hopefully at least it's clear.

**Review by npaton — CHANGES_REQUESTED** [review]
Good stuff 👍 
Couple of changes. And also one missing change in the answer.go file, that's another place we will add Participants, and I actually wonder if it might be the only place we add new Participants eventually, so we should probably have the same logic you have for `endMTurkHITStep()` in there too. Please let me know if my memory is wrong and something is not happening like I think it is. 👍

**Comment by hubertusputu1** [comment]
Yes, you are correct, we create another participant on `answers.go` file, i have already added it on there as well.

File: internal/runtime/process.go [hunk 5]
```diff
@@ -254,6 +251,11 @@ func (r *runState) filterParticipants(ctx context.Context, tx *ent.Tx, limit int
 		participants = participants[:n]
 	}
 
+	if uninitialized != true {
+		l := math.Min(float64(limit), float64(len(participants)))
+		return participants[:int(l)], nil
+	}
```
Comment (npaton) [inline internal/runtime/process.go:257]: `if !unitilized` (don't need to compare a boolean) you don't know filter out uninitialized?

I think all that's needed is to have a condition that handles both opposite cases in the loop below:

```go
if (!uninitialized && participant.Uninitialized != nil && *participant.Uninitialized) || (uninitialized && (participant.Uninitialized == nil || !*participant.Uninitialized))
```

**Review by npaton — CHANGES_REQUESTED** [review]
I don't think the participant filter is capturing the logic yet, see inline comment.

**Review by npaton — APPROVED** [review]

**Pull request merged**